### ISSUE TITLE
[question] how to set request-specific data in extensions

### ISSUE THREAD

**Issue by onichandame** [async-graphql/async-graphql#778]
### Edited

In the `prepare_request` method of an extension, the request is passed in as an argument. Hence I naiively thought that the data passed in the request would be propagated over. For example:

```rust
async fn prepare_request(
    &self,
    ctx: &ExtensionContext<'_>,
    request: Request,
    next: NextPrepareRequest<'_>,
) -> ServerResult<Request> {
    // The code here will be un before the prepare_request is executed, just after the request lifecycle hook.
    let mut request = request;
    request = request.data(String::from("hello")); // pass new data to request
    let result = next.run(ctx, request).await;
    // The code here will be run just after the prepare_request
    result
}
```

However, the new data is not actually propagated to the next extension or the resolver.

By looking into [the code](https://github.com/async-graphql/async-graphql/blob/master/src/extensions/mod.rs#L181), the `session_data` in the request is not merged to the context, so the next extension cannot access the new data through `Context`. The same happens for [the resolver](https://github.com/async-graphql/async-graphql/blob/master/src/schema.rs#L430)

My question is, is there a recommended way to pass data down the extension chain? If not, what is the consideration behind?

---
### Original

The [docs of extension](https://async-graphql.github.io/async-graphql/en/extensions.html) is lacking many important examples. Here is one question that cannot be answered by the docs: how to pass a new data down the extension chain?

For example, the request method is defined like:
```rust
async fn request(&self, ctx: &ExtensionContext<'_>, next: NextRequest<'_>) -> Response {
    next.run(ctx).await
}
```

Now suppose I am pre-processing the request in this method and wanting to pass the result to the next extensions and finally the resolver. How can I achieve it?

**Comment by arkus7** [comment]
I've encountered the same issue 😢 Is there any way to include some additional data in the extension, so it's usable in the handler itself?

**Comment by Miaxos** [comment]
Hello! If you want to access to the Context from inside your extensions, you'll just have to execute something like that:

```rust
#[async_graphql::async_trait::async_trait]
impl Extension for LoggerExtension {
    async fn prepare_request(
        &self,
        ctx: &ExtensionContext<'_>,
        request: Request,
        next: NextPrepareRequest<'_>,
    ) -> ServerResult<Request> {
        match ctx.data::<String>() {
            Ok(data) => println!("Good: {}", &data),
            Err(_) => println!("Bad"),
        };
        next.run(ctx, request).await
    }
}
```

**Comment by onichandame** [comment]
> Hello!
> 
> > My question is, is there a recommended way to pass data down the extension chain? If not, what is the consideration behind?
> 
> Yes the context :-)
> 
> If you want to access to the Context from inside your extensions, you'll just have to execute something like that:
> 
> ```rust
> #[async_graphql::async_trait::async_trait]
> impl Extension for LoggerExtension {
>     async fn prepare_request(
>         &self,
>         ctx: &ExtensionContext<'_>,
>         request: Request,
>         next: NextPrepareRequest<'_>,
>     ) -> ServerResult<Request> {
>         match ctx.data::<String>() {
>             Ok(data) => println!("Good: {}", &data),
>             Err(_) => println!("Bad"),
>         };
>         next.run(ctx, request).await
>     }
> }
> ```

This approach seems to only give access to the data passed in the context before the extensions are mounted/executed, e.g. from the schema build phase.

may i clarify my question better, the extensions are run in the same order they are mounted. i want to pass a data to the context from the former extension and access the data in the latter extension.

**Comment by rhuffus** [comment]
I have the same problem trying to add data in the prepare_request method of an Extension and trying after to get that data in the resolvers. Did you find any solution? Thanks.

**Comment by onichandame** [comment]
@rhuffus Sadly no. I gave up the extensions. For my specific use case, I need to decode the session from http requests and get the logged in user. What I do now is passing the session to the execution in the http handler. Then in resolver I manually decode the session. It is not the best way but AFAIK it is the only solution for now.

**Comment by sunli829** [comment]
> I have the same problem trying to add data in the prepare_request method of an Extension and trying after to get that data in the resolvers. Did you find any solution? Thanks.

Fixed in #1018， but no released yet.

**Comment by sunli829** [comment]
Released in `v4.0.10`
